Extensions/soft_refresh.js

```diff
@@ -11,212 +11,184 @@ XKit.extensions.soft_refresh = new Object({
-					var exists = $("[data-post-id='" + post_id + "']").length > 0;
-					exists = exists || $("[data-xkit-post-id='" + post_id + "']").length > 0;
-
-					if (!exists) {
-
-						m_count++;
-						post_ids.push(post_id);
-						if ($(".new_post_buttons_container").length > 0) {
-							$(".new_post_buttons_container").after($(this).parent()[0].outerHTML);
+					$("#posts > li", responseText)
+						.not("#new_post_buttons")
+						.not(".standalone-ad-container")
+						.each(function() {
+							const $this = $(this);
+							let exists = false;
+
+							if ($this.find("[data-sponsored], [data-is_recommended]").length) {
+								return;
+							}
+
+							if ($this.attr("data-pageable") !== undefined) {
+								exists = !!$("[data-pageable=" + $this.attr("data-pageable") + "]").length;
+								if (!exists) {
+									XKit.extensions.soft_refresh.post_ids.unshift($this.attr("data-pageable").replace("post_", ""));
+								}
+							}
+
+							if (!exists) {
+								XKit.extensions.soft_refresh.top_post.before($this);
+							} else {
+								end = true;
+								return false;
+							}
+						});
+
+					if (!end) {
+						this.request(`/2/${this.post_ids[0]}`);
+					} else {
+						if (this.post_ids.length === 0) {
+							if (this.preferences.show_notifications.value) {
+								XKit.notifications.add("No new posts found.", "info");
+							}
 						} else {
-							$("#new_post").after($(this)[0].outerHTML);
+							XKit.tools.add_function(this.hit_triggers, true);
+							if (this.preferences.show_notifications.value) {
+								XKit.notifications.add(`Added ${this.post_ids.length} new ${(this.post_ids.length === 1 ? "post" : "posts")}.`, "ok");
+							}
 						}
 
+						$("#xkit_soft_refresh").slideUp("fast", function() { $(this).remove(); });
+						this.post_ids = [];
+						this.loading = false;
 					}
-
 				});
-				$("html, body").animate({ scrollTop: 0 }, "slow");
-				XKit.extensions.soft_refresh.do_post_ids();
-
-				if (m_count === 0) {
-					if (XKit.extensions.soft_refresh.preferences.show_notifications.value === true) {
-						XKit.notifications.add("No new posts found.", "info");
-					}
-				} else {
-					XKit.tools.add_function(soft_refresh_hit_triggers, true, post_ids);
-					XKit.extensions.soft_refresh.check_embeds();
-					if (XKit.extensions.soft_refresh.preferences.show_notifications.value === true) {
-						XKit.notifications.add("Added " + m_count + " new posts.", "ok");
-					}
-				}
-				XKit.extensions.soft_refresh.loading = false;
-			}
-		});
-
+			})
+			.catch(() => this.show_error());
 	},
 
-	check_embeds: function() {
-
-		if ($("body").find(".inline_embed").length > 0) {
-			$("body").find(".inline_embed").each(function() {
-
-				try {
-					var script = document.createElement("script");
-					script.textContent = $(this).html();
-					document.body.appendChild(script);
-					$(this).remove();
-				} catch (e) {
-					console.error(e.message);
-				}
+	hit_triggers: function() {
```

Comment:
let's call this something a little more descriptive, like `notify_page` or `notify_tumblr` or `trigger_tumblr_events`